src/platform/backends/qemu/qemu_virtual_machine.cpp

```diff
@@ -260,6 +260,8 @@ void mp::QemuVirtualMachine::on_error()
 void mp::QemuVirtualMachine::on_shutdown()
 {
     state = State::off;
+    ip = nullopt;
+    dnsmasq_server->release_mac(mac_addr);
```

Comment:
As I mentioned in a comment, but maybe it would be better to do the explicit release when an instance is purged.